src/commands/doctor.ts

```diff
@@ -0,0 +1,151 @@
+
+  // Hide for now
+  public static hidden = true;
+
+  protected static flagsConfig = {
+    command: flags.string({
+      char: 'c',
+      description: messages.getMessage('flags.command'),
+    }),
+    plugin: flags.string({
+      char: 'p',
+      description: messages.getMessage('flags.plugin'),
+    }),
+    outputdir: flags.directory({
+      char: 'o',
+      description: messages.getMessage('flags.outputdir'),
+    }),
+  };
+
+  // Array of promises that are various doctor tasks to perform
+  // such as running a command and running diagnostics.
+  private tasks: Array<Promise<void>> = [];
+  private doctor: SfDoctor;
+  private outputDir: string;
+  private filesWrittenMsgs: string[] = [];
+
+  public async run(): Promise<SfDoctorDiagnosis> {
+    this.doctor = SFDoctor.getInstance();
+    const lifecycle = Lifecycle.getInstance();
+
+    const pluginFlag = this.flags.plugin as string;
+    const commandFlag = this.flags.command as string;
+    const outputdirFlag = this.flags.outputdir as string;
+    this.outputDir = path.resolve(outputdirFlag ?? process.cwd());
+
+    // eslint-disable-next-line @typescript-eslint/require-await
+    lifecycle.on<DiagnosticStatus>('Doctor:diagnostic', async (data) => {
+      this.ux.log(`${data.status} - ${data.testName}`);
+    });
+
+    if (commandFlag) {
+      this.setupCommandExecution(commandFlag);
+    }
+
+    if (pluginFlag) {
+      // verify the plugin flag matches an installed plugin
+      if (!this.config.plugins.some((p) => p.name === pluginFlag)) {
+        throw new SfError(messages.getMessage('pluginNotInstalledError', [pluginFlag]));
+      }
+
+      // run the diagnostics for a specific plugin
+      this.ux.styledHeader(`Running diagnostics for plugin: ${pluginFlag}`);
+      this.tasks.push(lifecycle.emit(`sf-doctor-${pluginFlag}`, this.doctor));
+    } else {
+      this.ux.styledHeader('Running all diagnostics');
+      // run all diagnostics
+      this.tasks.push(lifecycle.emit('sf-doctor', this.doctor));
+
+      /* eslint-disable @typescript-eslint/ban-ts-comment,@typescript-eslint/no-unsafe-assignment */
+      // @ts-ignore Seems like a TypeScript bug. Compiler thinks doctor.diagnose() returns `void`.
+      this.tasks = [...this.tasks, ...this.doctor.diagnose()];
+      /* eslint-enable @typescript-eslint/ban-ts-comment,@typescript-eslint/no-unsafe-assignment */
+    }
+
+    await Promise.all(this.tasks);
+
+    const diagnosis = this.doctor.getDiagnosis();
+    const diagnosisLocation = this.doctor.writeFileSync(
+      path.join(this.outputDir, 'diagnosis.json'),
+      JSON.stringify(diagnosis, null, 2)
+    );
+    this.filesWrittenMsgs.push(`Wrote doctor diagnosis to: ${diagnosisLocation}`);
+
+    this.ux.log();
+    this.filesWrittenMsgs.forEach((msg) => this.ux.log(msg));
+
+    this.ux.log();
+    this.ux.styledHeader('Suggestions');
+    diagnosis.suggestions.forEach((s) => this.ux.log(`  * ${s}`));
+
+    return diagnosis;
```

Comment:
The return should be enhanced to include more than just the diagnosis.  E.g., all the test results and the locations of the command debug logs.